dashboard/agent.py

```diff
@@ -0,0 +1,207 @@
+import argparse
+import asyncio
+import logging
+import os
+import sys
+import traceback
+
+import aiohttp
+import aioredis
+from grpc.experimental import aio as aiogrpc
+
+import ray
+import ray.new_dashboard.consts as dashboard_consts
+import ray.new_dashboard.utils as dashboard_utils
+import ray.ray_constants as ray_constants
+import ray.services
+import ray.utils
+import psutil
+
+logger = logging.getLogger(__name__)
+
+aiogrpc.init_grpc_aio()
+
+
+class DashboardAgent(object):
+    def __init__(self,
+                 redis_address,
+                 redis_password=None,
+                 temp_dir=None,
+                 log_dir=None,
+                 node_manager_port=None,
+                 object_store_name=None,
+                 raylet_name=None):
+        """Initialize the DashboardAgent object."""
+        self._agent_cls_list = dashboard_utils.get_all_modules(
+            dashboard_consts.TYPE_AGENT)
+        ip, port = redis_address.split(":")
+        # Public attributes are accessible for all agent modules.
+        self.redis_address = (ip, int(port))
+        self.redis_password = redis_password
+        self.temp_dir = temp_dir
+        self.log_dir = log_dir
+        self.node_manager_port = node_manager_port
+        self.object_store_name = object_store_name
+        self.raylet_name = raylet_name
+        self.ip = ray.services.get_node_ip_address()
+        self.server = aiogrpc.server(options=(("grpc.so_reuseport", 0), ))
+        listen_address = "[::]:0"
+        logger.info("Dashboard agent listen at: %s", listen_address)
+        self.port = self.server.add_insecure_port(listen_address)
+        self.aioredis_client = None
+        self.aiogrpc_raylet_channel = aiogrpc.insecure_channel("{}:{}".format(
+            self.ip, self.node_manager_port))
+        self.http_session = aiohttp.ClientSession(
+            loop=asyncio.get_event_loop())
+        self.metric_exporter = dashboard_utils.MetricExporter.create(
+            "OpenTSDBMetricExporter", http_session=self.http_session)
+
+    def _load_modules(self):
```

Comment:
So, the module exists because we'd like to have a set of async classes that run while loops & we'd like to simplify setting them up right? This means each module "should" have an async function that contains a function `run`? In this case, why don't we have an `ABC` class (like DashboardModule) that requires each module to implement some interface?

Reply:
The decorator `@agent` and `@master` in utils.py will check the interface of module, and it has an option `enable`. One decorator do all the things.

Another way is to make the module inherit from an `ABC` class.

I agree with you, inherit from an `ABC` class is better. I will fix this problem.